pkg/graph/resolvers/system_intake_document.go

```diff
@@ -84,9 +99,115 @@ func CreateSystemIntakeDocument(ctx context.Context, store *storage.Store, s3Cli
+// DeleteSystemIntakeDocument deletes an existing SystemIntakeDocument, given its ID.
 //
 // Does *not* delete the uploaded file from S3, following the example of TRB request documents.
 func DeleteSystemIntakeDocument(ctx context.Context, store *storage.Store, id uuid.UUID) (*models.SystemIntakeDocument, error) {
+	if err := canDelete(ctx, store, id); err != nil {
+		return nil, err
+	}
+
 	return store.DeleteSystemIntakeDocument(ctx, id)
 }
+
+// canView guards unauthorized views (downloads) of system intake documents
+// admins can view any document
+// requesters can view any document
+// GRB reviewers can view any document
+func canView(ctx context.Context, store *storage.Store, s3Key string) error {
+	// admins can view
+	if services.HasRole(ctx, models.RoleEasiGovteam) {
+		return nil
+	}
+
+	document, err := store.GetSystemIntakeDocumentByS3Key(ctx, s3Key)
+	if err != nil {
+
+		if errors.Is(err, sql.ErrNoRows) {
+			// if the document was deleted, don't error and break
+			return nil
+		}
+
+		return err
+	}
+
+	// requester can view
+	user := appcontext.Principal(ctx).Account()
+	if document.GetCreatedBy() == user.Username {
+		return nil
+	}
+
+	// if needed, get GRB members for this intake
+	grbUsers, err := store.SystemIntakeGRBReviewersBySystemIntakeIDs(ctx, []uuid.UUID{document.SystemIntakeRequestID})
+	if err != nil {
+		return err
+	}
+
+	if isGRBViewer := slices.ContainsFunc(grbUsers, func(reviewer *models.SystemIntakeGRBReviewer) bool {
+		return reviewer.UserID == user.ID
+	}); isGRBViewer {
+		return nil
+	}
+
+	appcontext.ZLogger(ctx).Warn("unauthorized user attempted to view system intake document",
+		zap.String("username", user.Username),
+		zap.String("system_intake.id", document.SystemIntakeRequestID.String()),
+		zap.String("document.id", document.ID.String()))
+
+	return errors.New("unauthorized attempt to view system intake document")
+}
+
+// canCreate guards unauthorized creation of system intake documents
+// admins can upload documents
+// requesters can upload documents
+// this also returns the role to use for the uploader, if authorized
+func canCreate(ctx context.Context, store *storage.Store, systemIntakeID uuid.UUID) (models.DocumentUploaderRole, error) {
+	intake, err := store.FetchSystemIntakeByID(ctx, systemIntakeID)
+	if err != nil {
+		return "", err
+	}
+
+	// check requester first as that role takes precedence over admin role for uploader roles
+	user := appcontext.Principal(ctx).Account()
+	if intake.EUAUserID.String == user.Username {
+		return models.RequesterUploaderRole, nil
+	}
+
+	if services.HasRole(ctx, models.RoleEasiGovteam) {
+		return models.AdminUploaderRole, nil
+	}
+
+	appcontext.ZLogger(ctx).Warn("unauthorized user attempted to create system intake document",
+		zap.String("username", user.Username),
+		zap.String("system_intake.id", systemIntakeID.String()))
+
+	return "", errors.New("unauthorized attempt to create system intake document")
+}
```

Comment:
I almost think this should be `getUploaderRole` instead of `canCreate` and then either check the role in the function that calls this or create a separate `canCreate` helper to throw the error. Thoughts?

```suggestion
func getUploaderRole(ctx context.Context, store *storage.Store, systemIntakeID uuid.UUID) (models.DocumentUploaderRole, error) {
	intake, err := store.FetchSystemIntakeByID(ctx, systemIntakeID)
	if err != nil {
		return "", err
	}

	// check requester first as that role takes precedence over admin role for uploader roles
	user := appcontext.Principal(ctx).Account()
	if intake.EUAUserID.String == user.Username {
		return models.RequesterUploaderRole, nil
	}

	if services.HasRole(ctx, models.RoleEasiGovteam) {
		return models.AdminUploaderRole, nil
	}

	return "", nil
}
```

Reply:
agree on naming and will change - will just remove `canCreate` naming for now

Reply:
changed naming to `validateUploader`

Reply:
i still want this to be specific for uploading a doc (and to make it make sense to return an unauthorized error)